File: smokers/smokers.c
#include <pthread.h>
#include <semaphore.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>

#define INGREDIENTS 3

enum INGREDIENT {
    PAPER = 0,
    TOBACCO = 1,
    MATCHES = 2
};

sem_t sem_table, sem_smoker[INGREDIENTS];

void smoke_cigarette(int index) {
    switch (index) {
        case PAPER:
            printf("[PAPER - smoker %02d] Smoking cigarette ... \n", index);
            break;
        case TOBACCO:
            printf("[TOBACCO - smoker %02d] Smoking cigarette ... \n", index);
            break;
        case MATCHES:
            printf("[MATCHES - smoker %02d] Smoking cigarette ... \n", index);
            break;
    }
    sleep(3);
}

void *agent(void *data) {
    while (1) {
        sem_wait(&sem_table);
        switch (rand() % INGREDIENTS)
        {
        case PAPER:
            printf("[agent] Puts tobacco and matches\n");
            sem_post(&sem_smoker[PAPER]);
            break;
        case TOBACCO:
            printf("[agent] Puts paper and matches\n");
            sem_post(&sem_smoker[TOBACCO]);
            break;
        case MATCHES:
            printf("[agent] Puts paper and tobacco\n");
            sem_post(&sem_smoker[MATCHES]);
            break;        
        default:
            break;
        } 
    }
}

void *smoker(void *data) {
    while (1) {
        int index = (int) data;
        sem_wait(&sem_smoker[index]);
        smoke_cigarette(index);
        sem_post(&sem_table);
    }
}

int main()
{
    pthread_t agent_th, smoker_th[INGREDIENTS];

    srand(time(NULL));

    sem_init(&sem_table, 0, 1);
    for (int i = 0; i<INGREDIENTS; i++) {
        sem_init(&sem_smoker[i], 0, 0);
    }

    pthread_create(&agent_th, NULL, agent, NULL);
    for (int i = 0; i<INGREDIENTS; i++) {
        pthread_create(&smoker_th[i], NULL, smoker, (void *) i);
    }

    pthread_join(agent_th, NULL);
    for (int i = 0; i<INGREDIENTS; i++) {
        pthread_join(smoker_th[i], NULL);
    }

    sem_destroy(&sem_table);
    for (int i = 0; i<INGREDIENTS; i++) {
        sem_destroy(&sem_smoker[i]);
    }
}

File: smokers/smokers.py
import logging
import threading
import time
import random

def thread_agent(name):
    logging.info("Agent starting")
    while True:
        s_table.acquire()
        gen = random.randint(0,2)
        if gen == 0:
            s_tobacco.release()
            s_matches.release()
            logging.info("Puts tobacco and matches")
        elif gen == 1:
            s_paper.release()
            s_matches.release()
            logging.info("Puts paper and matches")
        else:
            s_paper.release()
            s_tobacco.release()
            logging.info("Puts tobacco and paper")
            
    logging.info("Agent finishing")

def obtain_ingredients(name, s1, s2):
    s1.acquire()
    if s2.acquire(blocking=False):
        logging.info("%s smoking", name)
        time.sleep(2)
        s_table.release()
    else:
        s1.release()

def thread_smoker(id, name):
    logging.info("Thread %s (%s): starting", name, id)
    while True:
        if id == 0:
            obtain_ingredients(name, s_tobacco, s_matches)
        elif id == 1:
            obtain_ingredients(name, s_paper, s_matches)
        else:
            obtain_ingredients(name, s_tobacco, s_paper)


ingredients = [ "PAPER", "TOBACCO", "MATCHES" ]

s_table = threading.Semaphore(1)
s_paper = threading.Semaphore(0)
s_tobacco = threading.Semaphore(0)
s_matches = threading.Semaphore(0)

if __name__ == "__main__":
    format = "%(asctime)s: %(message)s"
    logging.basicConfig(format=format, level=logging.INFO,
                        datefmt="%H:%M:%S")

    agent = threading.Thread(target=thread_agent, args=("agent",))
    agent.start()

    threads = list()
    for index in range(3):
        s = threading.Thread(target=thread_smoker, args=(index, ingredients[index],))
        threads.append(s)
        s.start()

    for index, thread in enumerate(threads):
        thread.join()
    
    agent.join()